relayer/relays/beacon/header/syncer/syncer.go

```diff
@@ -493,55 +523,127 @@ func (s *Syncer) unmarshalBeaconState(slot uint64, data []byte) (state.BeaconSta
 		if err != nil {
-			slot -= s.setting.SlotsInEpoch
+			if highestSlot < slot || head.Slot < slot {
+				return 0, fmt.Errorf("unable to find valid slot")
+			}
+
+			slot += s.protocol.Settings.SlotsInEpoch
 
 			continue
 		}
 
-		headers = append(headers, header.Slot)
-		headers = append(headers, finalizedHeader.Slot)
+		log.WithFields(log.Fields{"attested": attestedSlot, "finalized": finalizedSlot}).Info("found boundary headers")
+		return attestedSlot, nil
+	}
+}
+
+func (s *Syncer) findValidUpdatePair(slot uint64) (uint64, uint64, error) {
+	finalizedHeader, err := s.Client.GetHeaderBySlot(slot)
+	if err != nil {
+		return 0, 0, fmt.Errorf("get finalized slot: %d err: %w", slot, err)
+	}
+
+	attestedSlot := finalizedHeader.Slot + (s.protocol.Settings.SlotsInEpoch * 2)
+	attestedHeader, err := s.Client.GetHeaderBySlot(attestedSlot)
+	if err != nil {
+		return 0, 0, fmt.Errorf("get attested slot: %d err: %w", attestedSlot, err)
+	}
+
+	nextHeader, err := s.FindBeaconHeaderWithBlockIncluded(attestedSlot + 1)
+	if err != nil {
+		return 0, 0, fmt.Errorf("get next header: %d err: %w", attestedSlot+1, err)
+	}
+	nextBlock, err := s.Client.GetBeaconBlockBySlot(nextHeader.Slot)
+	if err != nil {
+		return 0, 0, fmt.Errorf("get next block: %d err: %w", nextHeader.Slot, err)
+	}
+
+	superMajority, err := s.protocol.SyncCommitteeSuperMajority(nextBlock.Data.Message.Body.SyncAggregate.SyncCommitteeBits)
+	if err != nil {
+		return 0, 0, fmt.Errorf("compute sync committee supermajority: %d err: %w", nextHeader.Slot, err)
+	}
+	if !superMajority {
+		return 0, 0, fmt.Errorf("sync committee at slot not supermajority: %d", nextHeader.Slot)
 	}
 
-	log.WithField("headers_found", headers).Info("found boundary headers")
-	return headers[0], nil
+	return finalizedHeader.Slot, attestedHeader.Slot, nil
 }
 
-func (s *Syncer) GetFinalizedUpdateAtAttestedSlot(attestedSlot uint64, lastSyncedFinalizedSlot uint64) (scale.Update, error) {
-	var update scale.Update
+func (s *Syncer) GetLatestPossibleFinalizedUpdate(attestedSlot uint64, boundary uint64) (scale.Update, error) {
+	attestedSlot, err := s.FindLatestAttestedHeadersAtInterval(attestedSlot, boundary)
+	if err != nil {
+		return scale.Update{}, fmt.Errorf("cannot find blocks at boundaries: %w", err)
+	}
+
+	return s.GetFinalizedUpdateAtAttestedSlot(attestedSlot, boundary, false)
+}
+
+func (s *Syncer) GetFinalizedUpdateWithSyncCommittee(syncCommitteePeriod, lastFinalizedSlot uint64) (scale.Update, error) {
+	slot := (syncCommitteePeriod) * s.protocol.Settings.SlotsInEpoch * s.protocol.Settings.EpochsPerSyncCommitteePeriod
+
+	// If the latest finalized slot is equal or larger than the slot to be synced, the finalized header
+	// won't be imported and so it should be a newer header. 3 epochs further, because the finalized header is
+	// 2 epochs behind the attested header, and we need 1 epoch newer at least
+	if lastFinalizedSlot >= slot {
+		slot += s.protocol.Settings.SlotsInEpoch * 3
+	}
```

Comment:
Seems not correct,  I'd assume the logic here is to advance `slot` by at least 3 epochs or more until `slot > lastFinalizedSlot`?

Reply:
I added this because I was worried that the latest finalized header would be larger than the update with the sync committee. But I think it is not a problem, since an older finalized update is allowed if the update contains a sync committee period (and NextSyncCommittee is not set): https://github.com/Snowfork/polkadot-sdk/blob/snowbridge/bridges/snowbridge/pallets/ethereum-client/src/lib.rs#L330

So I will remove this check.

Reply:
> The latest finalized header would be larger than the update with the sync committee

I think it's normal in that case we just sync committee without updating finalized state
https://github.com/Snowfork/polkadot-sdk/blob/93687126c499f40d71baba83567936d56434dac0/bridges/snowbridge/pallets/ethereum-client/src/lib.rs#L465-L467

Reply:
Removed in https://github.com/Snowfork/snowbridge/pull/1164/commits/9cbee1db2231aed977210ccd0fbdbea17d412cf8.